client/client-lib/src/wallet/mod.rs

```diff
@@ -210,6 +211,10 @@ mod tests {
         async fn register_gateway(&self, _gateway: LightningGateway) -> crate::api::Result<()> {
             unimplemented!()
         }
+
+        async fn fetch_epoch_history(&self, _epoch: u64) -> crate::api::Result<EpochHistory> {
+            unimplemented!()
+        }
```

Comment:
Is there any way we can define a default impl of `FederationApi` -- at least in tests? It's kind of annoying how we always need to create all these `unimplemented!` methods ...

Reply:
Agreed, outside of macros though I'm not sure the best was to handle this since Rust doesn't have inheritance.

Reply:
Inheritance is evil. :)

Just a thought but maybe the interface is just too big here. "Interface segregation principle".

In Rust that could be:

```rust
trait Foo {}

trait Bar {}

trait Ban {}

trait FullApi : Foo + Bar + Ban {}

trait FooBarApi : Foo + Bar {}

type OwnedFooBarApi = Box<dyn FooBarApi>;
type OwnedFullApi = Box<dyn FooBarApi>;
```

As you can see unfortunately  `Box<dyn Boo + Bar>` doesn't work, and it requires creating explicit `trait` for every needed combination. And there might be other downsides that I can't see RN, so I would not be overdoing it. But if you're finding yourself adding a lot of `unimplemented!` in fake impls, you might to think a bit and split thing into some logically separate pieces.

Reply:
Haha yeah I don't think Rust needs inheritance in general, but having something like mix-in or default impls would be nice to have sometimes.  Probably the interface or more likely the test fixture needs to be refactored, but it'd do it outside this PR.

Reply:
Traits to have default implementations. You could just add `unimplemented!();` everywhere by default. Not sure if it's a best practice (might lead to minor mistakes in the prod impls, especially if there's more of them), but it would work.

Alternatively you can do some trait + generics sheningans. Let me come up with something ...

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=00b9f5361f2a75aa95322162bba7b159

I'm afraid I've invented a mocking framework here, possibly better ones already exist. :D

Reply:
> You could just add `unimplemented!();` everywhere by default

Agreed sounds like a dangerous idea.  As far as mocking libraries go we could probably find something, although we'd need to be able to implement functions rather than just returning mock values.

Reply:
> As far as mocking libraries go we could probably find something, although we'd need to be able to implement functions rather than just returning mock values.

Can you explain?

Reply:
Currently for testing we actually implement some functionality, so I'm not sure whether a mock framework would be sufficient.

Anyway, I'm going to move this work to a new issue https://github.com/fedimint/fedimint/issues/369